In [1]:
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt
from tensorflow import keras
import random
import pandas as pd

tf.logging.set_verbosity(tf.logging.INFO)

!pip install librosa
import librosa

In [0]:
s, sr=librosa.load('train_clean_male.wav', sr=None)
S_sgn=librosa.stft(s, n_fft=1024, hop_length=512)

sn, sr=librosa.load('train_dirty_male.wav', sr=None)
X_sgn=librosa.stft(sn, n_fft=1024, hop_length=512)

In [3]:

S_sgn = np.abs(S_sgn.T)
S_sgn = S_sgn
X_sgn = np.abs(X_sgn.T)
X_sgn = X_sgn
print(X_sgn.shape)

(2459, 513)


**1-D CNN  structure**


*   For Audio cleaning problem a 1-d convolutional neural network is built with the following:
* Input is 513 shape vector 
* 2 convolutional layers:
  - One with 32 filters, filter size of (3,3)
  - One with 64 filters, filter size of (2,2)
* Each convolution layer is followed by a max-pooling layer with pooling size (2,2) and stride taken as 2 units.
* 1 dense fully connected layer with 2048 and Relu as activation function
* Dropout of 0.4 on dense layer
* Logits layer to output with 513 neurons to match output shape



In [0]:
lr = 1e-3
n_iterations = 1000
batch_size = 128
dropout = 0.5

n_input = 513   # input layer (28x28 pixels)
n_dense = 2048
n_output = 513 
filter1 = 32
kernel_size1 = 3
filter2 = 64
kernel_size2 = 2
drop_out = 0.4

In [0]:
# def cnn_model_fn(features, labels, mode):
  """Model function for CNN."""
  # Input Layer
# X = tf.reshape(X_sgn, [-1, 512, 1])
# Y = tf.reshape(S_sgn, [-1, 512, 1])
X = tf.placeholder("float", [None, n_input])
Y = tf.placeholder("float", [None, n_output])


X1 = tf.reshape(X, [-1, n_input, 1])
Y1 = tf.reshape(Y, [-1, n_output, 1])

# Convolutional Layer #1
conv1 = tf.layers.conv1d(
    inputs=X1,
    filters=filter1,
    kernel_size=(kernel_size1),
    padding="same",
    activation=tf.nn.relu)

# Pooling Layer #1
pool1 = tf.layers.max_pooling1d(inputs=conv1, pool_size=[2], strides=2)

# Convolutional Layer #2 and Pooling Layer #2
conv2 = tf.layers.conv1d(
    inputs=pool1,
    filters=filter2,
    kernel_size=(kernel_size2),
    padding="same",
    activation=tf.nn.relu)
pool2 = tf.layers.max_pooling1d(inputs=conv2, pool_size=[2], strides=2)
print(pool2.shape)
# Dense Layer
pool2_flat = tf.reshape(pool2, [-1, 128 * filter2])
dense = tf.layers.dense(inputs=pool2_flat, units=n_dense, activation=tf.nn.relu)
dropout = tf.layers.dropout(
    inputs=dense, rate=drop_out)

# Logits Layer
logits = tf.layers.dense(inputs=dense, units=n_output)
print(logits.shape)
# cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(labels=Y, logits=output_layer))
loss=tf.losses.mean_squared_error(labels=tf.reshape(Y1, [-1, n_output]), predictions=logits)
train_step = tf.train.AdamOptimizer(lr).minimize(loss)
# print(logits)

# print(S_sgn)
correct_pred = tf.equal(tf.argmax(logits, 1), tf.argmax(Y1, 1))
accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32))
# print(accuracy)

In [0]:
init = tf.global_variables_initializer()
sess = tf.Session()
sess.run(init)

** Mini Batching **
* For mini batches, the input is randomly sampled.
* A batch that is based on the batch size given is taken from the input at a random position.
* Each batch and corresponding labels are fed to the training model at once.

In [0]:
for i in range(n_iterations):
  idx = random.randint(0,20)
  batch_x = X_sgn[idx*batch_size:idx*batch_size + batch_size,:]
  batch_y = S_sgn[idx*batch_size:idx*batch_size + batch_size,:]
  sess.run(train_step, feed_dict={X: batch_x, Y: batch_y})
#   sess.run(train_step, feed_dict={X: X_sgn[0,:], Y: S_sgn[0,:]})
  print(i)

** Testing **
* Method to process test files and create out put .wav files.

In [0]:
# Method to refine test audio files
def process_test(file, out):

  test, sr = librosa.load(file, sr=None)
  test_sgn = librosa.stft(test, n_fft=1024, hop_length=512)
  print(test_sgn.shape)
  magnitude, phase = librosa.magphase(test_sgn)
  test_s = np.abs(test_sgn)

  pred = sess.run(logits, feed_dict={X: test_s.T})
  pred = np.multiply(np.divide(test_sgn,test_s),pred.T)
#   print(pred.shape)
  sh_test = librosa.istft(pred, hop_length=512, length=len(test))
#   print("SNR :")
#   print(sh_test.shape)
#   print(snr(test, sh_test))
  if(out == 1):
    outfile = 'test_s_01_recons.wav'
  else:
    outfile = 'test_s_02_recons.wav'
  librosa.output.write_wav(outfile, sh_test, sr)
  
  
def snr(signal, est):
  num = np.sum(np.square(signal))
  den = np.sum(np.square(signal-est))
  snr = 10* np.log10(num/den)
  return snr

** SNR **
* Method to process train file get predictions and find SNR to evaluate the model.

In [0]:
def process_train(file, out):

  test, sr = librosa.load(file, sr=None)
  test_sgn = librosa.stft(test, n_fft=1024, hop_length=512)
  test_s = np.abs(test_sgn)

  pred = sess.run(logits, feed_dict={X: test_s.T})
  pred = np.multiply(np.divide(test_sgn,test_s),pred.T)
  sh_test = librosa.istft(pred, hop_length=512, length=len(test))
  
  
  s, sr = librosa.load('train_clean_male.wav', sr=None)
  print(snr(s, sh_test))
  outfile= "train_s_recons.wav"
  librosa.output.write_wav(outfile, sh_test, sr)
  return snr(s, sh_test)

In [17]:
# process_test('test_x_01.wav',1)
# process_test('test_x_02.wav',2)
print('SNR for train sample :')
snr = process_train('train_dirty_male.wav',1)


SNR for train sample :
16.940776109695435


In [0]:
with open("performance.txt", "a") as text_file:
    print(f"CNN 1", file=text_file)
    print(f"SNR: {snr}", file=text_file)
#     print(f"LR: {lr}", file=text_file)
    print(f"LR: {lr}\t Iterations: {n_iterations}\t Batch Size: {batch_size}", file=text_file)
    print(f"Filter1: {filter1}\t Kernel1: {kernel_size1}", file=text_file)
    print(f"Filter2: {filter2}\t Kernel2: {kernel_size2}", file=text_file)
    print(f"Dense neurons: {n_dense}\t Drop out: {drop_out}\n", file=text_file)

** Result **
* With given parameters for 1-d CNN the SNR for train file is 16.94. \
**SNR**: **16.940776109695435** \
**Learning Rate**: 0.001	 **Iterations**: 1000	\
**Batch Size**: 128 \
**Filter1**: 32	 **Kernel1**: 3 \
**Filter2**: 64	 **Kernel2**: 2 \
**Dense neurons**: 2048	 \
Drop out: 0.4
* The output files generated for test audio samples seem reasonable with less noise of chips but the performance did not seem as good as that of fully connected layer as far as audio is concerned.